File: app/src/main/java/net/xzos/UpgradeAll/server/app/engine/js/utils/JSUtils.java
package net.xzos.UpgradeAll.server.app.engine.js.utils;

import net.xzos.UpgradeAll.application.MyApplication;
import net.xzos.UpgradeAll.gson.JSCacheData;
import net.xzos.UpgradeAll.server.log.LogUtil;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.jsoup.Connection;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.seimicrawler.xpath.JXDocument;
import org.seimicrawler.xpath.JXNode;

import java.util.ArrayList;

/**
 * 爬虫相关库的打包集合
 * For JavaScript
 */
public class JSUtils {

    private static final LogUtil Log = MyApplication.getServerContainer().getLog();
    private static final String TAG = "JSUtils";
    private String[] logObjectTag;

    private JSCacheData jsCacheData = new JSCacheData();

    public JSUtils(String[] logObjectTag) {
        this.logObjectTag = logObjectTag;
    }

    JSCacheData getJsCacheData() {
        return jsCacheData;
    }

    public void setJsCacheData(JSCacheData jsCacheData) {
        this.jsCacheData = jsCacheData;
    }


    public JSONObject getJSONObject() {
        return new JSONObject();
    }

    public JSONArray getJSONArray() {
        return new JSONArray();
    }

    public JSONObject getJSONObject(String jsonString) throws JSONException {
        return new JSONObject(jsonString);
    }

    public JSONArray getJSONArray(String jsonString) throws JSONException {
        return new JSONArray(jsonString);

    }

    public String getHttpResponse(String URL) {
        JSONObject httpResponseDict = jsCacheData.getHttpResponseDict();
        String responseString = null;
        if (httpResponseDict.has(URL)) {
            try {
                responseString = httpResponseDict.getString(URL);
                Log.d(logObjectTag, TAG, "getHttpResponse: 从缓存加载, URL: " + URL);
            } catch (JSONException e) {
                Log.e(logObjectTag, TAG, "getHttpResponse: HTTP 缓存队列无该对象, httpResponseDict : " + httpResponseDict);
            }
        } else {
            responseString = OkHttpApi.getHttpResponse(logObjectTag, URL);
            if (responseString != null) {
                try {
                    httpResponseDict.put(URL, responseString);
                    Log.d(logObjectTag, TAG, "getHttpResponse: 缓存, URL: " + URL);
                } catch (JSONException e) {
                    Log.d(logObjectTag, TAG, "getHttpResponse: 缓存失败, URL: " + URL);
                }
            }
        }
        return responseString;
    }

    public ArrayList selNByJsoupXpath(String userAgent, String URL, String xpath) {
        Document doc = new Document(URL);
        JSONObject jsoupDomDict = jsCacheData.getJsoupDomDict();
        if (jsoupDomDict.has(URL)) {
            try {
                doc = (Document) jsoupDomDict.get(URL);
                Log.d(logObjectTag, TAG, "selNByJsoupXpathJavaList: 从缓存加载, URL: " + URL);
            } catch (JSONException e) {
                Log.e(logObjectTag, TAG, "selNByJsoupXpathJavaList: Jsoup 缓存队列无该对象, jsoupDomDict: " + jsoupDomDict);
            }
        } else {
            Connection connection = Jsoup.connect(URL);
            if (userAgent != null) connection.userAgent(userAgent);
            doc = JsoupApi.getDoc(connection);
            if (doc == null) {
                Log.e(logObjectTag, TAG, "selNByJsoupXpathJavaList: Jsoup 对象初始化失败");
                return new ArrayList<>();
            }
            try {
                jsoupDomDict.put(URL, doc);
                Log.d(logObjectTag, TAG, "selNByJsoupXpathJavaList: 缓存, URL: " + URL);
            } catch (JSONException e) {
                Log.d(logObjectTag, TAG, "selNByJsoupXpathJavaList: 缓存失败, URL: " + URL);
            }
        }
        JXDocument JXDoc = JXDocument.create(doc);
        ArrayList<String> nodeStringArrayList = new ArrayList<>();
        for (JXNode node : JXDoc.selN((xpath))) {
            nodeStringArrayList.add(node.toString());
        }
        Log.d(logObjectTag, TAG, "selNByJsoupXpath: node_list number: " + nodeStringArrayList.size());
        return nodeStringArrayList;
    }
}

File: app/src/main/kotlin/net/xzos/upgradeAll/server/app/engine/js/utils/JSUtils.kt
package net.xzos.upgradeAll.server.app.engine.js.utils

import android.widget.Toast
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import net.xzos.upgradeAll.application.MyApplication.Companion.context
import net.xzos.upgradeAll.data.json.nongson.JSCache
import net.xzos.upgradeAll.data.json.nongson.MyCookieManager
import net.xzos.upgradeAll.data.json.nongson.ObjectTag
import net.xzos.upgradeAll.server.log.LogUtil
import net.xzos.upgradeAll.utils.MiscellaneousUtils
import net.xzos.upgradeAll.utils.VersioningUtils
import net.xzos.upgradeAll.utils.network.AriaDownloader
import net.xzos.upgradeAll.utils.network.JsoupApi
import net.xzos.upgradeAll.utils.network.OkHttpApi
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.mozilla.javascript.Context
import org.seimicrawler.xpath.JXDocument


/**
 * 爬虫相关库的打包集合
 * For JavaScript
 */
internal class JSUtils(
        private val objectTag: ObjectTag,
        internal var debugMode: Boolean = false,
        jsCache: JSCache = JSCache(objectTag)
) {

    private lateinit var cx: Context

    private val jsoupApi = JsoupApi(objectTag, jsCache)
    private val okHttpApi = OkHttpApi(objectTag, jsCache)

    fun get(cx: Context): JSUtils = this.also {
        this.cx = cx
    }

    fun parseJSONObject(): JSONObject {
        return JSONObject()
    }

    fun parseJSONArray(): JSONArray {
        return JSONArray()
    }

    @Throws(JSONException::class)
    fun parseJSONObject(jsonString: String? = null): JSONObject {
        return if (jsonString != null) JSONObject(jsonString)
        else JSONObject()
    }

    @Throws(JSONException::class)
    fun parseJSONArray(jsonString: String? = null): JSONArray {
        return if (jsonString != null) JSONArray(jsonString)
        else JSONArray()
    }

    fun getHttpResponse(URL: String): String? =
            okHttpApi.getHttpResponse(URL, catchError = false)

    fun selNByJsoupXpath(userAgent: String?, URL: String, xpath: String): MutableList<String> {
        val doc = jsoupApi.getDoc(URL, userAgent = userAgent)
                ?: return mutableListOf()
        val jxDocument = JXDocument.create(doc)
        val nodeStringList = mutableListOf<String>().apply {
            for (node in jxDocument.selN(xpath)) {
                this.add(node.toString())
            }
        }
        Log.d(objectTag, TAG, "selNByJsoupXpath: node_list number: " + nodeStringList.size)
        return nodeStringList
    }

    fun matchVersioningString(versionString: String?): String? =
            VersioningUtils.matchVersioningString(versionString)

    fun downloadFile(fileName: String, URL: String, headers: Map<String, String> = mapOf(),
                     isDebug: Boolean = debugMode, externalDownloader: Boolean = false) {
        val allHeaders = hashMapOf<String, String>().apply {
            this.putAll(headers)
            this.putAll(jsoupApi.requestHeaders) // 装载由 Jsoup 生成的正常 header
            this.putAll(okHttpApi.requestHeaders) // 装载由 Jsoup 生成的正常 header
            MyCookieManager.getCookiesString(URL)?.let {
                this["Cookie"] = it
            }  // 装载 Cookies
        }
        val ariaDownloader = AriaDownloader(isDebug).apply {
            createDownloadTaskNotification(fileName)
        }
        val resUrl = jsoupApi.getRedirectsUrl(URL, headers = allHeaders) ?: URL
        if (!externalDownloader)
            try {
                ariaDownloader.start(
                        fileName, resUrl,
                        headers = allHeaders)
            } catch (e: IllegalArgumentException) {
                Log.e(objectTag, TAG, """ downloadFile: 下载任务失败
                        |下载参数: URL: $resUrl, FileName: $fileName, headers: $allHeaders
                        |ERROR_MESSAGE: $e""".trimIndent())
                ariaDownloader.cancel()
                runBlocking(Dispatchers.Main) {
                    Toast.makeText(context, "下载失败: $fileName", Toast.LENGTH_SHORT).show()
                }
            }
        else {
            MiscellaneousUtils.accessByBrowser(
                    resUrl,
                    context
            )
            ariaDownloader.cancel()
        }
    }

    companion object {
        private val Log = LogUtil
        private const val TAG = "JSUtils"
    }
}